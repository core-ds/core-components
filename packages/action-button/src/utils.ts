import { getDataTestId } from '@alfalab/core-components-shared';

export function getActionButtonTestIds(dataTestId: string) {
    return {
        actionButton: dataTestId,
        spinner: getDataTestId(dataTestId, 'loader'),
    };
}
