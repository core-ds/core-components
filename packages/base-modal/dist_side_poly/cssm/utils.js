'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var coreComponentsGlobalStore = require('@alfalab/core-components-global-store/cssm');
var coreComponentsShared = require('@alfalab/core-components-shared/cssm');

function isScrolledToTop(target) {
    return target.scrollTop <= 0;
}
function isScrolledToBottom(target) {
    return target.scrollHeight - target.offsetHeight <= target.scrollTop;
}
function hasScrollbar(target) {
    return target.scrollHeight > target.clientHeight;
}
var isOverflowing = function (container) {
    if (document.body === container) {
        return window.innerWidth > document.documentElement.clientWidth;
    }
    return container.scrollHeight > container.clientHeight;
};
var getPaddingRight = function (node) {
    return parseInt(window.getComputedStyle(node).paddingRight, 10) || 0;
};
var restoreContainerStyles = function (container) {
    var modalRestoreStyles = coreComponentsGlobalStore.getModalStore().getRestoreStyles();
    var index = modalRestoreStyles.findIndex(function (s) { return s.container === container; });
    var existingStyles = modalRestoreStyles[index];
    if (!existingStyles)
        return;
    existingStyles.modals -= 1;
    if (existingStyles.modals <= 0) {
        modalRestoreStyles.splice(index, 1);
        existingStyles.styles.forEach(function (_a) {
            var value = _a.value, el = _a.el, key = _a.key;
            if (value) {
                el.style.setProperty(key, value);
            }
            else {
                el.style.removeProperty(key);
            }
        });
    }
};
var handleContainer = function (container, shouldIOSLock) {
    if (shouldIOSLock === void 0) { shouldIOSLock = false; }
    if (!container)
        return;
    var modalRestoreStyles = coreComponentsGlobalStore.getModalStore().getRestoreStyles();
    var existingStyles = modalRestoreStyles.find(function (s) { return s.container === container; });
    if (existingStyles) {
        existingStyles.modals += 1;
        return;
    }
    var containerStyles = [];
    if (isOverflowing(container)) {
        // Вычисляет размер до применения `overflow hidden` для избежания скачков
        var scrollbarSize = coreComponentsShared.browser.getScrollbarSize();
        containerStyles.push({
            value: container.style.paddingRight,
            key: 'padding-right',
            el: container,
        });
        // Вычисляем стили, чтобы получить реальный `padding` c шириной сколлбара
        // eslint-disable-next-line no-param-reassign
        container.style.paddingRight = "".concat(getPaddingRight(container) + scrollbarSize, "px");
    }
    var parent = container.parentElement;
    var scrollContainer = 
    // TODO: заменить на optional chaining
    parent &&
        parent.nodeName === 'HTML' &&
        window.getComputedStyle(parent).overflowY === 'scroll'
        ? parent
        : container;
    // Блокируем скролл даже если отсутствует скроллбар
    if (scrollContainer.style.overflow !== 'hidden') {
        containerStyles.push({
            value: scrollContainer.style.overflow,
            key: 'overflow',
            el: scrollContainer,
        });
    }
    if (!shouldIOSLock) {
        scrollContainer.style.overflow = 'hidden';
    }
    modalRestoreStyles.push({
        container: container,
        modals: 1,
        styles: containerStyles,
    });
};

exports.handleContainer = handleContainer;
exports.hasScrollbar = hasScrollbar;
exports.isScrolledToBottom = isScrolledToBottom;
exports.isScrolledToTop = isScrolledToTop;
exports.restoreContainerStyles = restoreContainerStyles;
