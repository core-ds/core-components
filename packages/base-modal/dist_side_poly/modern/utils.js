import { getModalStore } from '@alfalab/core-components-global-store/modern';
import { browser } from '@alfalab/core-components-shared/modern';

function isScrolledToTop(target) {
    return target.scrollTop <= 0;
}
function isScrolledToBottom(target) {
    return target.scrollHeight - target.offsetHeight <= target.scrollTop;
}
function hasScrollbar(target) {
    return target.scrollHeight > target.clientHeight;
}
const isOverflowing = (container) => {
    if (document.body === container) {
        return window.innerWidth > document.documentElement.clientWidth;
    }
    return container.scrollHeight > container.clientHeight;
};
const getPaddingRight = (node) => parseInt(window.getComputedStyle(node).paddingRight, 10) || 0;
const restoreContainerStyles = (container) => {
    const modalRestoreStyles = getModalStore().getRestoreStyles();
    const index = modalRestoreStyles.findIndex((s) => s.container === container);
    const existingStyles = modalRestoreStyles[index];
    if (!existingStyles)
        return;
    existingStyles.modals -= 1;
    if (existingStyles.modals <= 0) {
        modalRestoreStyles.splice(index, 1);
        existingStyles.styles.forEach(({ value, el, key }) => {
            if (value) {
                el.style.setProperty(key, value);
            }
            else {
                el.style.removeProperty(key);
            }
        });
    }
};
const handleContainer = (container, shouldIOSLock = false) => {
    if (!container)
        return;
    const modalRestoreStyles = getModalStore().getRestoreStyles();
    const existingStyles = modalRestoreStyles.find((s) => s.container === container);
    if (existingStyles) {
        existingStyles.modals += 1;
        return;
    }
    const containerStyles = [];
    if (isOverflowing(container)) {
        // Вычисляет размер до применения `overflow hidden` для избежания скачков
        const scrollbarSize = browser.getScrollbarSize();
        containerStyles.push({
            value: container.style.paddingRight,
            key: 'padding-right',
            el: container,
        });
        // Вычисляем стили, чтобы получить реальный `padding` c шириной сколлбара
        // eslint-disable-next-line no-param-reassign
        container.style.paddingRight = `${getPaddingRight(container) + scrollbarSize}px`;
    }
    const parent = container.parentElement;
    const scrollContainer = 
    // TODO: заменить на optional chaining
    parent &&
        parent.nodeName === 'HTML' &&
        window.getComputedStyle(parent).overflowY === 'scroll'
        ? parent
        : container;
    // Блокируем скролл даже если отсутствует скроллбар
    if (scrollContainer.style.overflow !== 'hidden') {
        containerStyles.push({
            value: scrollContainer.style.overflow,
            key: 'overflow',
            el: scrollContainer,
        });
    }
    if (!shouldIOSLock) {
        scrollContainer.style.overflow = 'hidden';
    }
    modalRestoreStyles.push({
        container,
        modals: 1,
        styles: containerStyles,
    });
};

export { handleContainer, hasScrollbar, isScrolledToBottom, isScrolledToTop, restoreContainerStyles };
