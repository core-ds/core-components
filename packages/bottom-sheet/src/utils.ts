export const TIMEOUT = 500;
export const SWIPE_VELOCITY = 0.4;
export const MARKER_HEIGHT = 24;
/* Верхний отступ шторки, если она открыта на максимальную высоту */
export const HEADER_OFFSET = 24;
export const CLOSE_OFFSET = 0.2;
export const SCROLL_OFFSET = 1;

export const convertPercentToNumber = (value: string | number, fullHeight: number) => {
    const maxHeight = fullHeight - HEADER_OFFSET;

    if (typeof value === 'string') {
        const percent = parseFloat(value) / 100;

        if (percent < 0) {
            return Math.min(maxHeight, fullHeight + fullHeight * percent);
        }

        return Math.min(maxHeight, fullHeight * percent);
    }

    if (value < 0) {
        return Math.max(0, fullHeight + value);
    }

    return Math.min(maxHeight, value);
};
