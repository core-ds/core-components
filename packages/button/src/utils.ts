import { getDataTestId } from '@alfalab/core-components-shared';

export function getButtonTestIds(dataTestId: string) {
    return {
        button: dataTestId,
        spinner: getDataTestId(dataTestId, 'loader'),
    };
}
