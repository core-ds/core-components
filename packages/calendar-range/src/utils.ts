import { getDataTestId } from '@alfalab/core-components-shared';

export const isDayButton = (node: HTMLElement | null) =>
    node && node.tagName === 'BUTTON' && node.dataset.date;

export function getCalendarRangeTestIds(dataTestId: string) {
    return {
        component: dataTestId,
        calendarContainerFrom: getDataTestId(dataTestId, 'container-from'),
        calendarContainerTo: getDataTestId(dataTestId, 'container-to'),
    };
}
