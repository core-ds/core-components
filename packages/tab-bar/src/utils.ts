import { getDataTestId } from '@alfalab/core-components-shared';

export function getTabBarTestIds(dataTestId: string) {
    return {
        tabBar: dataTestId,
        tab: getDataTestId(dataTestId, 'tab'),
    };
}
