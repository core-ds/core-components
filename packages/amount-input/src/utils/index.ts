/**
 * https://github.com/core-ds/utils
 * TODO: нужно перенести утилиты в этот проект, когда он будет готов
 */

import { CurrencyCodes } from '@alfalab/data';
import { formatAmount } from '@alfalab/utils';

/**
 * Форматирует введенное значение
 * @param enteredValue Значение введенное в инпут
 * @param currency валюта
 * @param minority количество минорных единиц
 */
export function getFormattedValue(enteredValue: string, currency: CurrencyCodes, minority: number) {
    if (!enteredValue || enteredValue === '-') {
        return enteredValue;
    }

    // eslint-disable-next-line prefer-const
    let [head, tail] = enteredValue.split(',');

    // При вводе "-," указываем, что имеется в виду "-0,"
    if (head === '-') {
        head = '-0';
    }

    let { majorPart } = formatAmount({
        value: Number(head) * minority,
        currency,
        minority,
        negativeSymbol: 'hyphen-minus',
    });

    // Так как -0 === 0, formatAmount возвращает положительное значение. Исправляем это здесь
    if (/^-(,|0),?/.test(enteredValue) && majorPart === '0') {
        majorPart = `-${majorPart}`;
    }

    if (!tail && enteredValue.includes(',')) {
        return majorPart.concat(',');
    }

    if (tail) {
        return majorPart.concat(',', tail.slice(0, minority.toString().length - 1));
    }

    return majorPart;
}

export function getAmountValueFromStr(str: string, minority: number) {
    if (str === '' || str === '-') {
        return null;
    }

    return Math.round(Number(str.replace(',', '.').replace(/[^0-9.-]/g, '')) * minority);
}
