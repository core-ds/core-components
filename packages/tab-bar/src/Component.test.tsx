import { fireEvent, render } from '@testing-library/react';
import React, { forwardRef } from 'react';
import { getTabBarTestIds } from './utils';
import { TabBar, TabBarProps, TabProps } from '.';

const dti = 'tab-bar';

const TabBarComponent = ({
    tabProps,
    ...restProps
}: Partial<TabBarProps> & { tabProps?: Partial<TabProps> }) => {
    const [selectedId, setSelectedId] = React.useState('1');

    return (
        <TabBar selectedId={selectedId} onChange={setSelectedId} dataTestId={dti} {...restProps}>
            <TabBar.Tab id='1' label='TabName1' icon='icon1' {...tabProps} />
            <TabBar.Tab id='2' label='TabName2' icon='icon2' {...tabProps} />
        </TabBar>
    );
};

describe('TagBar', () => {
    describe('snapshot test', () => {
        it('should match snapshot', () => {
            const { container } = render(<TabBarComponent />);
            expect(container).toMatchSnapshot();
        });
    });

    describe('attributes tests', () => {
        it('should set `data-test-id` attribute', () => {
            const { container, getAllByTestId } = render(<TabBarComponent />);

            const testIds = getTabBarTestIds(dti);

            expect(getAllByTestId(testIds.tabBar).length).toBe(1);
            expect(getAllByTestId(testIds.tab).length).toBe(2);
        });

        it('should set `className` attribute', () => {
            const { getByTestId } = render(<TabBarComponent className='tab-bar' />);

            expect(getByTestId(dti)).toHaveClass('tab-bar');
        });

        it('should set `border` className if border={true} ', () => {
            const { getByTestId } = render(<TabBarComponent border={true} />);

            expect(getByTestId(dti)).toHaveClass('border');
        });

        it('should set `modal-bg-alt-primary` className if bgColor=`modal-bg-alt-primary`', () => {
            const { getByTestId } = render(<TabBarComponent bgColor='modal-bg-alt-primary' />);

            expect(getByTestId(dti)).toHaveClass('modal-bg-alt-primary');
        });

        it('should set `secondary` className if accentColor=`secondary`', () => {
            const { getAllByTestId } = render(<TabBarComponent accentColor='secondary' />);

            expect(getAllByTestId(`${dti}-tab`)[0]).toHaveClass('secondary');
        });

        it('should set classes for every tab if tabClassNames specified', () => {
            const { getAllByTestId } = render(
                <TabBarComponent
                    tabClassNames={{
                        className: 'tab-class',
                        iconClassName: 'icon-class',
                        selectedClassName: 'selected-class',
                        labelClassName: 'label-class',
                    }}
                />,
            );

            getAllByTestId(`${dti}-tab`).forEach((tab) => expect(tab).toHaveClass('tab-class'));
            getAllByTestId(`${dti}-tab`).forEach((tab) => {
                const icon = tab.querySelector('.icon');
                expect(icon).toHaveClass('icon-class');
            });
            getAllByTestId(`${dti}-tab`).forEach((tab) => {
                const label = tab.querySelector('.label');
                expect(label).toHaveClass('label-class');
            });
            expect(getAllByTestId(`${dti}-tab`)[0]).toHaveClass('selected-class');
            expect(getAllByTestId(`${dti}-tab`)[1]).not.toHaveClass('selected-class');
        });

        it('should set `className` attribute for tab', () => {
            const { container } = render(<TabBarComponent tabProps={{ className: 'tab' }} />);

            expect(container.querySelectorAll('.tab').length).toBe(2);
        });

        it('should set `iconClassName` attribute for tab', () => {
            const { container } = render(
                <TabBarComponent tabProps={{ iconClassName: 'icon-class' }} />,
            );

            expect(container.querySelectorAll('.icon-class').length).toBe(2);
        });

        it('should set `labelClassName` attribute for tab', () => {
            const { container } = render(
                <TabBarComponent tabProps={{ labelClassName: 'label-class' }} />,
            );

            expect(container.querySelectorAll('.label-class').length).toBe(2);
        });

        it('should set `selectedClassName` attribute for tab', () => {
            const { container } = render(
                <TabBarComponent tabProps={{ selectedClassName: 'selected-class' }} />,
            );

            expect(container.querySelectorAll('.selected-class').length).toBe(1);
        });

        it('should set `id` attribute for tab', () => {
            const { container } = render(<TabBarComponent tabProps={{ id: 'tab-id' }} />);

            expect(container.querySelectorAll('#tab-id').length).toBe(2);
        });
    });

    describe('render tests', () => {
        it('should render icon component', () => {
            const Icon = () => <span>icon</span>;

            const { container } = render(<TabBarComponent tabProps={{ icon: <Icon /> }} />);

            container.querySelectorAll('.icon').forEach((icon) => {
                expect(icon).toHaveTextContent('icon');
            });
        });

        it('should render label component', () => {
            const Label = () => <span>label</span>;

            const { container } = render(<TabBarComponent tabProps={{ label: <Label /> }} />);

            container.querySelectorAll('.label').forEach((label) => {
                expect(label).toHaveTextContent('label');
            });
        });

        it('should render indicator', () => {
            const { container } = render(<TabBarComponent tabProps={{ showIndicator: true }} />);

            expect(container.querySelectorAll('.indicator').length).toBe(2);
        });

        it('should use custom component', () => {
            const cb = jest.fn();
            cb.mockReturnValue(null);
            const to = { path: '/test' };

            render(<TabBarComponent tabProps={{ Component: forwardRef(cb), to }} />);

            expect(cb).toBeCalled();

            const props = cb.mock.calls[0][0];
            expect(props.to).toBe(to);
        });
    });

    describe('callback tests', () => {
        it('should call `onChange` callback', () => {
            const onChange = jest.fn();
            const { getAllByRole } = render(<TabBarComponent onChange={onChange} />);

            const tab2 = getAllByRole('button')[1];
            fireEvent.click(tab2);

            expect(onChange).toBeCalledTimes(1);
            expect(onChange).toBeCalledWith('2');
        });
    });
});
