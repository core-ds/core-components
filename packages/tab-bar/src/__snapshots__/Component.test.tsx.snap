// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagBar snapshot test should match snapshot 1`] = `
<div>
  <nav
    class="component modal-bg-primary"
    data-test-id="tab-bar"
  >
    <button
      class="component primary"
      data-test-id="tab-bar-tab"
      id="1"
      tabindex="0"
      type="button"
    >
      <div
        class="icon"
      >
        icon1
      </div>
      <span
        class="label labelSelected secondary-small medium"
      >
        TabName1
      </span>
    </button>
    <button
      class="component"
      data-test-id="tab-bar-tab"
      id="2"
      tabindex="0"
      type="button"
    >
      <div
        class="icon"
      >
        icon2
      </div>
      <span
        class="label secondary-small medium"
      >
        TabName2
      </span>
    </button>
  </nav>
</div>
`;
